reconciler/reconcile_common.go

```diff
@@ -29,12 +30,20 @@ const failedGenerationBump = "NewObservedGenFailure"
 func PreProcessReconcile(ctx context.Context, resource duckv1.KRShaped) {
 	newStatus := resource.GetStatus()
 
-	if newStatus.ObservedGeneration != resource.GetGeneration() {
-		condSet := resource.GetConditionSet()
-		manager := condSet.Manage(newStatus)
-		// Ensure conditions are initialized before we modify.
-		manager.InitializeConditions()
+	// We may be reading a version of the object that was stored at an older version
+	// and may not have had all of the assumed defaults specified.  This won't result
+	// in this getting written back to the API Server, but lets downstream logic make
+	// assumptions about defaulting.
+	if d, ok := resource.(apis.Defaultable); ok {
+		d.SetDefaults(ctx)
```

Comment:
This might be a problem for Autoscaler, because those are pluggable and can be a third party implementation. The implementation is chosen based on the annotation. And one of the important debugging signals here is uninitialized conditions if this annotations is set incorrectly.
Now we do filter (https://github.com/knative/serving/blob/master/pkg/reconciler/autoscaling/kpa/controller.go#L101) those, so I _think_ this should be alright, but want to double check that this won't default if we're not actually going into the `ReconcileKind`.

Reply:
This is _only_ called immediately before `ReconcileKind` https://github.com/knative/pkg/blob/22662525751b007a039ef72c05193f6a0d3602ee/codegen/cmd/injection-gen/generators/reconciler_reconciler.go#L315